### DISCUSSION TITLE
Calculate principal curvature directions

### DISCUSSION THREAD

**Discussion by jhenriques** [pmp-library/pmp-library#220] [Q&A] [answered]
Hi all,

How can I use the curvature (or the gradient_matrix()) data to calculate the principal curvature directions (the vectors).

I would expect to be able to use these to have vertex bi-normals. Is this possible?

**Comment by mbotsch** [comment, marked as answer]
We currently don't provide computation of principal curvature directions. Might be an interesting feature addition though.

**Reply by jhenriques** [reply]
If I wanted to implement this myself, do you have any pointers/ideas on how this would be done?

**Reply by mbotsch** [reply]
If I'm not mistaken, the principal curvature directions should be the eigenvectors of the symmetric 3x3 tensor in line 229 of the file curvature.cpp. It even seems that we already compute the eigenvectors (line 232). The eigenvectors have to be sorted accordingly to the sorting of the eigenvalues (the block starting at line 235). Then it should be the two eigenvectors corresponding to the two larger eigenvalues. The eigenvalue corresponding to the smallest eigenvalue should be (roughly) the normal vector.

Please give it a try and let me know if it doesn't work.
